server/server.go

```diff
@@ -194,33 +233,32 @@ func (s *RPCServer) consume(queueName string, handler handlerFunc, inputCh *amqp
 		return err
 	}
 
+	// attach the middlewares to the handler.
+	handler = Middlewares(handler, s.middlewares...)
+
 	go func() {
 		logger.Infof("server: waiting for messages on queue '%s'", queue.Name)
 
 		for delivery := range deliveries {
 			logger.Infof("server: got delivery on queue %v correlation id %v", queue.Name, delivery.CorrelationId)
-			var (
-				errMiddleware error
-				response      []byte
-			)
-
-			for _, middleware := range s.middlewares {
-				errMiddleware = middleware(queue.Name, context.TODO(), &delivery)
-				if errMiddleware != nil {
-					response = []byte(errMiddleware.Error())
-					break
-				}
-			}
 
-			if errMiddleware == nil {
-				response = handler(context.TODO(), delivery)
+			rw := ResponseWriter{
+				publishing: &amqp.Publishing{
+					CorrelationId: delivery.CorrelationId,
+					Body:          []byte{},
+				},
 			}
 
+			ctx := context.WithValue(s.ctx, CtxQueueName, queue.Name)
+
+			handler(ctx, &rw, delivery)
 			delivery.Ack(false)
 
 			s.responses <- processedRequest{
-				response: response,
-				delivery: delivery,
+				replyTo:    delivery.ReplyTo,
+				mandatory:  rw.mandatory,
```

Comment:
Three values passed form the `ResponseWriter` then used right away. Any good reasons to not just put the response writer in the `processedRequest`?

Reply:
Mostly because I want to release the `ResponseWriter` as soon as possible, so that calling any method on it after the handler has returned will have no affect.